### PR TITLE
`v2.1` Feature: Replicates Instead of Cross Validation Folds

### PR DESCRIPTION
**Pull request by JacksonBurns**
## Description
This PR is a follow up to an offline meeting, where we discussed changing Chemprop to use replicates instead of cross validation.

## Example / Current workflow
Chemprop's `--num-folds` arguments controls how many cross-validation folds to perform, and behind the scenes Chemprop does something similar to the outer loop of nested cross validation, meaning that all of train, validation, and test are 'striped' through the data simultaneously. The size of each set is determined by the number of folds.

## Bugfix / Desired workflow
This PR changes Chemprop to instead do __replicates__, i.e. independent re-samplings. Advantages of this approach include:
 - the 'real' performance of a model can typically be approximated using fewer replicates than cross-validation folds, reducing the amount of training users need (i.e. 4 replicates of 80:10:10 splitting will give similar results to 8-fold cross validation with half the number of models)
 - literature is increasingly moving towards replicates, especially of splits other than random - see our peers like UniMol [here](https://openreview.net/pdf?id=6K2RM6wVqKu) using replicates of scaffold splitting and MHNN [here](https://arxiv.org/pdf/2312.13136) doing replicates of random splitting
 - this implementation of replicates works with all of the different splitting methods, allowing people to replicate random, kmeans, etc.
 - replicates avoid introducing accidental structure into the data, a theoretical possibility with cross-validation

## Checklist
- [x] linted with flake8?
- [x] (if appropriate) unit tests ~~added?~~ modified

### CODE REVIEW

**Comment by JacksonBurns** [comment]
Also, `tests/unit/data/test_data_utils.py` is failing right now because I just haven't updated those tests yet, and the regression tests are failing because changing this changes the expected results.

**Comment by shihchengli** [comment]
I think we can still keep the CV methods and make the `--num-replicates` to be `--n-splits` to match the sklearn styles. In v1, we do have both CV methods and replicates implemented. I don't think it would hurt to keep both. For the scaffold splits, how does Astartes handle cases where most of the data are non-scaffold? Would it raise an error if the number of scaffolds is too small?

I noticed that we implemented `cv-no-val` instead of `cv-no-test`. Is there a reason to implement `cv-no-val` but not `cv-no-test`? `cv-no-test` would allow the use of all the data for model training, while `cv-no-val` could lead to overfitting as there's no validation metric calculated. I think we should remove `cv-no-val` and implement `cv-no-test`.

**Comment by JacksonBurns** [comment]
If there are not enough scaffolds, resulting in validation or test being empty, `astartes` would raise an error about the empty set.

Agreed that we should remove no-val in favor of no-test - we can roll that into this PR.

**Comment by JacksonBurns** [comment]
TODO: add a demo notebook demonstrating how to do CV if people want to as a module

**Comment by JacksonBurns** [comment]
Update following our offline ChemProp developers meeting:
 - we will continue with removing cross-validation in favor of replicates in this PR
 - a demo notebook will be added to show how to use cross validation via scikit-learn's many cross validation classes
 - we will add descriptive errors in the CLI to move people towards the new arguments

**Comment by JacksonBurns** [comment]
@shihchengli @KnathanM this PR is ready for review.
 - `-k/--num-folds` now raises an Error and tells users to switch to `--num-replicates` - in 2.1 we will remove the former entirely
 - the tests and notebooks have been updated to use the new syntax (since our `make_split_indices` function now always returns the same type)
 - CV is gone entirely, as we discussed in our offline meeting

Please let me know your thoughts. We are ideally looking to merge after 2.0.5 and before 2.1.0, which would mean we have a 1-2 week window starting 1 week from now.

**Comment by JacksonBurns** [comment]
We will have to update the documentation notebooks before merging, but unit tests and demo notebooks will pass. I will update the former once the rest of this is approved.

**Comment by ankur-soley** [comment]
Is this feature available yet on any github branch? I am currently using `--num-folds 10` and my job dies just after first cross-validation step on Chemprop v2.0.4. It is woking fine in Chemprop v1.

**Comment by JacksonBurns** [comment]
@ankur-soley you can install this version of ChemProp with `pip install git+https://github.com/chemprop/chemprop.git@refs/pull/994/merge`, but your issue sounds unrelated to this. Please consider opening a new issue.

**Review by donerancl — CHANGES_REQUESTED** [review]
Other than my two comments on replacing `num_folds` with `num_replicates` in documentation `splitting.ipynb` notebook, and potentially raising a error when `num_folds` is not `None` for `make_split_indices`, it looks good to me.

File: chemprop/cli/train.py [hunk 1]
```diff
@@ -50,6 +50,9 @@
 logger = logging.getLogger(__name__)
 
 
+_CV_DEP_WARNING = "The -k/--num-folds argument is deprecated and will be completely removed in v2.2.0 - use --num-replicates instead."
```
Comment (KnathanM) [inline chemprop/cli/train.py:53]: A depreciation warning sounds like --num-folds should still work, but you set this message to be displayed with an argument error. I think it okay to give an error with this message when the flag is given in v.2.1 and then remove the error message and flag in v2.2, but is there a stronger word than "deprecated", something like:
```suggestion
_CV_DEP_WARNING = "The -k/--num-folds argument was removed in v2.1.0, please use --num-replicates instead."
```

**Review by KnathanM — COMMENTED** [review]
I don't love that with this change we have to do `train_data[0]` in the notebooks instead of just `train_data`, but I can't really argue with how much simpler it makes the splitting code, so it gets a "yes" from me.

The changes look great, do you have time to also update the documentation? The relevant parts I found are:
1. [docs](https://chemprop.readthedocs.io/en/main/tutorial/python/data/splitting.html), [code](https://github.com/chemprop/chemprop/blob/main/docs/source/tutorial/python/data/splitting.ipynb)
2. [docs](https://chemprop.readthedocs.io/en/main/tutorial/cli/hpopt.html#splitting), [code](https://github.com/chemprop/chemprop/blob/main/docs/source/tutorial/cli/hpopt.rst)
3. [docs](https://chemprop.readthedocs.io/en/main/tutorial/cli/train.html#cross-validation), [code](https://github.com/chemprop/chemprop/blob/main/docs/source/tutorial/cli/train.rst)

File: chemprop/cli/train.py [hunk 2]
```diff
@@ -446,6 +444,9 @@ def process_train_args(args: Namespace) -> Namespace:
     if args.config_path is None and args.data_path is None:
         raise ArgumentError(argument=None, message="Data path must be provided for training.")
 
+    if args.num_folds is not None:  # i.e. user-specified
```
[the remark this answers is not in the document]
Comment (JacksonBurns) [reply chemprop/cli/train.py:471]: Great suggestion - done.

File: chemprop/data/splitting.py [hunk 3]
```diff
@@ -32,8 +30,8 @@ def make_split_indices(
     split: SplitType | str = "random",
     sizes: tuple[float, float, float] = (0.8, 0.1, 0.1),
     seed: int = 0,
-    num_folds: int = 1,
-):
+    num_replicates: int = 1,
```
[the remark this answers is not in the document]
Comment (JacksonBurns) [reply chemprop/data/splitting.py:33]: I'll do this as part of the tutorial updates in @KnathanM 's review comments, thank for mentioning it here as well.

File: chemprop/cli/train.py (hunk 1 above)
Comment (JacksonBurns) [reply chemprop/cli/train.py]: Good point - did this in all places where relevant in c852b75

File: examples/training.ipynb [hunk 4]
```diff
@@ -306,24 +315,49 @@
...
+            "outputs": [],
+            "source": [
+                "from sklearn.model_selection import KFold\n",
+                "\n",
+                "k_splits = KFold(n_splits=5)\n",
+                "k_train_indices, k_val_indices, k_test_indices = [], [], []\n",
+                "for fold in k_splits.split(mols):\n",
+                "    k_train_indices.append(fold[0])\n",
+                "    k_val_indices.append([])\n",
+                "    k_test_indices.append(fold[1])"
```
[the remark this answers is not in the document]
Comment (JacksonBurns) [reply examples/training.ipynb:335]: Great addition, committing.

**Comment by JacksonBurns** [comment]
@donerancl @KnathanM I have updated the docs and addressed all other comments, please re-review.

**Review by donerancl — APPROVED** [review]
LGTM!

File: docs/source/tutorial/python/data/splitting.ipynb [hunk 5]
```diff
@@ -95,7 +95,11 @@
    "source": [
     "1. Collect the `rdkit.Chem.mol` objects for each datapoint. These are required for structure based splits.\n",
     "2. Generate the splitting indices.\n",
-    "3. Split the data using those indices."
+    "3. Split the data using those indices.\n",
+    "\n",
+    "The `make_split_indices` function includes a `num_replicates` argument to perform repeated splits (each with a different ranodm seed) with your sampler of choice.\n",
```
Comment (KnathanM) [inline docs/source/tutorial/python/data/splitting.ipynb:100]:
```suggestion
"The `make_split_indices` function includes a `num_replicates` argument to perform repeated splits (each with a different random seed) with your sampler of choice.\n",
```

**Review by KnathanM — APPROVED** [review]

File: docs/source/tutorial/python/data/splitting.ipynb (hunk 5 above)
Comment (JacksonBurns) [reply docs/source/tutorial/python/data/splitting.ipynb:100]: Thanks for catching
